File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-1.java
class Solution {
    public int characterReplacement(String s, int k) {
        HashMap<Character, Integer> windowMap = new HashMap<>();
        int out = 1;
        int left = 0;
        windowMap.put(s.charAt(0), 1);
        for (int i = 1; i < s.length(); i++) {
            windowMap.put(s.charAt(i), windowMap.getOrDefault(s.charAt(i), 0)+1);
            int windowSize = i - left + 1;
            int most_freq = Collections.max(windowMap.values());
            if (windowSize - most_freq <= k) {
                out = Math.max(out, windowSize);
            } else {
                windowMap.put(s.charAt(left), windowMap.get(s.charAt(left))-1 );
                if (windowMap.get(s.charAt(left)) == 0) {
                    windowMap.remove(s.charAt(left));
                }
                left ++;
            }
        }
        return out;
    }
}
/*
AAABABB
AAAAA
*/

File: Data Structures & Algorithms/house-robber/submission-1.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        if n <= 1:
            return nums[0]
        if n <= 2:
            return max(nums)

        dp = [0] * (n+1)

        dp[0] = 0
        dp[1] = nums[0]
        dp[2] = nums[1]
        for i in range(3, n+1):
            dp[i] = max(dp[i-2], dp[i-3]) + nums[i-1]
        return max(dp)